cmd/osctl/cmd/cluster.go

```diff
@@ -118,238 +100,109 @@ func create() (err error) {
-	var wg sync.WaitGroup
-
-	wg.Add(len(requests))
-
-	for _, req := range requests {
-		go func(req *node.Request) {
-			fmt.Println("creating node", req.Name)
-
-			if err = node.NewNode(clusterName, req); err != nil {
-				helpers.Fatalf("failed to create node: %w", err)
-			}
-
-			wg.Done()
-		}(req)
-	}
-
-	wg.Wait()
-
-	return nil
-}
-
-func destroy() error {
-	cli, err := client.NewEnvClient()
-	if err != nil {
+	if err = saveConfig(cluster); err != nil {
 		return err
 	}
 
-	filters := filters.NewArgs()
-	filters.Add("label", "talos.owned=true")
-	filters.Add("label", "talos.cluster.name="+clusterName)
-
-	containers, err := cli.ContainerList(context.Background(), types.ContainerListOptions{All: true, Filters: filters})
-	if err != nil {
-		return err
+	if !clusterWait {
+		return nil
 	}
 
-	var wg sync.WaitGroup
-
-	wg.Add(len(containers))
-
-	for _, container := range containers {
-		go func(container types.Container) {
-			fmt.Println("destroying node", container.Names[0][1:])
+	// Run cluster readiness checks
+	checkCtx, checkCtxCancel := context.WithTimeout(ctx, 20*time.Minute)
+	defer checkCtxCancel()
 
-			err := cli.ContainerRemove(context.Background(), container.ID, types.ContainerRemoveOptions{RemoveVolumes: true, Force: true})
-			if err != nil {
-				helpers.Fatalf("%+v", err)
-			}
+	clusterAccess := access.NewAdapter(cluster)
+	defer clusterAccess.Close() //nolint: errcheck
 
-			wg.Done()
-		}(container)
-	}
-
-	wg.Wait()
-
-	fmt.Println("destroying network", clusterName)
-
-	return destroyNetwork(cli)
+	return check.Wait(checkCtx, clusterAccess, check.DefaultClusterChecks(), check.StderrReporter())
 }
 
-func ensureImageExists(ctx context.Context, cli *client.Client, image string) error {
-	// In order to pull an image, the reference must be in canononical
-	// format (e.g. domain/repo/image:tag).
-	ref, err := reference.ParseNormalizedNamed(image)
-	if err != nil {
-		return err
-	}
-
-	image = ref.String()
-
-	// To filter the images, we need a familiar name and a tag
-	// (e.g. domain/repo/image:tag => repo/image:tag).
-	familiarName := reference.FamiliarName(ref)
-	tag := ""
-
-	if tagged, isTagged := ref.(reference.Tagged); isTagged {
-		tag = tagged.Tag()
-	}
-
-	filters := filters.NewArgs()
-	filters.Add("reference", familiarName+":"+tag)
-
-	images, err := cli.ImageList(ctx, types.ImageListOptions{Filters: filters})
+func destroy(ctx context.Context) error {
+	provisioner, err := docker.NewProvisioner(ctx)
 	if err != nil {
 		return err
 	}
 
-	if len(images) == 0 {
-		fmt.Println("downloading", image)
-
-		var reader io.ReadCloser
-
-		if reader, err = cli.ImagePull(ctx, image, types.ImagePullOptions{}); err != nil {
-			return err
-		}
+	defer provisioner.Close() //nolint: errcheck
 
-		// nolint: errcheck
-		defer reader.Close()
+	return provisioner.Destroy(ctx, fakeCluster{clusterName})
+}
 
-		if _, err = io.Copy(ioutil.Discard, reader); err != nil {
-			return err
-		}
-	}
+type fakeCluster struct {
```

Comment:
A little confused by this. Is this temporary? What makes it "fake", and why can't we have a "real" one?

Reply:
this is something I don't have elegant solution for so far. In order to destroy a cluster, I want to have a `Cluster` description, which makes sense for many types of clusters I think (e.g. for firecracker). with Docker our flow relies only on cluster name, as Docker has rich metadata API and we can find nodes, network only by cluster name. So to satisfy the interface, here I build a structure which implements `Cluster` interface, but provides only `ClusterName` property, probably I should make it less cryptic and move it some package, as it might be only Docker-specific.

Reply:
I removed `fakeCluster` and replaced with `Provisioner.Reflect` method which allows to reflect back `Cluster` object from running configuration. It's not completely implemented yet, but enough for `cluster destroy` to work properly. My future plan is to have more complete support for reflection, so that we could potentially scale existing cluster, or wait for readiness of already existing cluster. Either reflection, or dumping/loading `Cluster` to `yaml` (might be better to see which ones works better after firecracker).